oauth2/oauth2_auth_code_test.go

```diff
@@ -1605,6 +1606,22 @@ func TestAuthCodeWithDefaultStrategy(t *testing.T) {
 						}
 						assert.Len(t, allTokens, (1+nRefreshes)*2)
 					})
+
+					t.Run("graceful refresh count limit is respected when set", func(t *testing.T) {
+						reg.Config().MustSet(ctx, config.KeyRefreshTokenRotationGraceReuseCount, 3)
+						reg.Config().MustSet(ctx, config.KeyRefreshTokenRotationGracePeriod, "1m")
+						reg.Config().MustSet(ctx, config.KeyRefreshTokenLifespan, "1m")
+
+						token := issueTokens(t)
+						token.Expiry = time.Now().Add(-time.Hour * 24)
+
+						for range 3 {
+							_, err := conf.TokenSource(ctx, token).Token()
+							require.NoError(t, err)
+						}
+						_, err := conf.TokenSource(ctx, token).Token()
+						assert.Error(t, err, "Rotating a used refresh token is not possible after the limit is exceeded")
```

Comment:
Do we need to add additional tests to make sure token chains are revoked as well, or is this covered with other tests? We have a bunch of those tests for the grace period

Reply:
It would be - in this file - "token generations" - where we create multiple generations of tokens and then see if revoking one revokes all. I think it should just work, but I'm wondering if we should still add a test case specifically for this.

Reply:
Added a case now.